ext/sqlite3/database.c

```diff
@@ -12,6 +12,59 @@
 
 VALUE cSqlite3Database;
 
+static void
+close_or_discard_db(sqlite3RubyPtr ctx)
+{
+    if (ctx->db) {
+        int isReadonly = (ctx->flags & SQLITE_OPEN_READONLY);
+
+        if (isReadonly || ctx->owner == getpid()) {
+            // Ordinary close.
+            sqlite3_close_v2(ctx->db);
+        } else {
+            // This is an open connection carried across a fork().
+            // "Discard" it. See adr/2024-09-fork-safety.md
+            sqlite3_file *sfile;
+            int status;
+
+            rb_warning("An open writable sqlite database connection was inherited from a "
+                       "forked process and is being closed to prevent the risk of corruption. "
+                       "If possible, please close all writable sqlite database connections "
+                       "before forking.");
```

Comment:
Do we still need this warning if we're auto-closing stuff cross fork?  I'm OK with keeping the warning. IMO if you leave the connection open and fork then it's probably a bug in your app and we should let the end user know.

Reply:
Yeah, thanks for asking, this is a good question. I am worried that this change is going to break somebody's app ("Why is this database connection closed?") and they'll have to dig around to figure out what's going on. But I guess that's what release notes are for? I could go either way but tend to err towards overcommunication.

Reply:
> I could go either way but tend to err towards overcommunication.

💯

Reply:
If people get annoyed we can remove it later.

Reply:
See https://github.com/sparklemotion/sqlite3-ruby/pull/566